libr/core/linux_heap_glibc_64.c

```diff
@@ -0,0 +1,815 @@
+	while (next && next >= brk_start && next <= main_arena->top) {
+		PRINTF_BA ("0x%"PFMT64x, next);
+		while (double_free == UT64_MAX && next_tmp) {
+			r_core_read_at (core, next_tmp, (ut8 *)cnk, sizeof (RHeapChunk64));
+			next_tmp = cnk->fd;
+			if (next_root == next_tmp) {
+				double_free = next_root;
+				break;
+			}
+		}
+		r_core_read_at (core, next, (ut8 *)cnk, sizeof (RHeapChunk64));
+		next = cnk->fd;
+
+		if (cnk->prev_size > size || cnk->size > size) {
+			PRINT_RA (" Linked list corrupted\n");
+			PRINT_GA ("\n  }\n");
+			return 1;	
+		}
+
+		PRINTF_BA ("%s", next ? "->fd = " : "");
+		next_root = next_tmp = next;
+		if (double_free == next) {
+			PRINTF_RA ("0x%"PFMT64x, next);
+			PRINT_RA (" Double free detected\n");
+			PRINT_GA ("\n  }\n");
+			return -1;
+		}
+	}
+
+	if (next && (next < brk_start || next > main_arena->top)) {
+		PRINTF_RA ("0x%"PFMT64x, next);
+		PRINT_RA (" Linked list corrupted\n");
+		PRINT_GA ("\n  }\n");
+		return 1;
+	} else {		
+		PRINT_GA ("\n  }\n");
+		return -1;			
+	}
+}
+
+void print_heap_fastbin_64(RCore *core, ut64 m_arena, RHeap_MallocState64 *main_arena, const char *input) {
+	int i;
+	ut64 num_bin = UT64_MAX;
+	ut64 offset = sizeof (int) * 2;
+
+	switch (input[0]) {
+	case '\0': // dmhf
+		PRINT_YA ("fastbinY {\n");
+		for (i = 1; i <= NFASTBINS; i++) {
+			PRINTF_YA (" Fastbin %02d\n", i);
+			if (!print_single_linked_list_bin_64 (core, main_arena, m_arena, offset, i - 1)) {
+				PRINT_GA ("  Empty bin");
+				PRINT_BA ("  0x0\n");
+			}
+		}
+		PRINT_YA ("}\n");
+		break;
+	case ' ': // dmhf [bin_num]
+		num_bin = r_num_math (core->num, input + 1) - 1;
+		if (num_bin >= NFASTBINS) {
+			eprintf ("Error: 0 < bin <= %d\n", NFASTBINS);
+			break;
+		}
+		if (!print_single_linked_list_bin_64 (core, main_arena, m_arena, offset, num_bin)) {
+			PRINT_GA (" Empty bin");
+			PRINT_BA (" 0x0\n");
+		}
+		break;
+	}
+}
+
+void print_heap_graph_64(RCore *core, RHeap_MallocState64 *main_arena, ut64 *initial_brk) {
+	int w, h;
+	ut64 top_size = UT64_MAX;
+	w = r_cons_get_size (&h);
+	RConsCanvas *can = r_cons_canvas_new (w, h);
+	can->color = r_config_get_i (core->config, "scr.color");
+	RAGraph *g = r_agraph_new (can);
+	RANode *top = {0}, *chunk_node = {0}, *prev_node = {0};
+	RListIter *iter;
+	RDebugMap *map;
+	RHeapChunk64 *cnk = R_NEW0 (RHeapChunk64), *prev_c = R_NEW0 (RHeapChunk64);
+
+	ut64 next_chunk_ref, prev_chunk_ref, brk_start, brk_end, size_tmp;
+	char *top_title, *top_data, *node_title, *node_data;
+	bool first_node = true;
+
+	r_agraph_set_title (g, "Heap Layout");
+	top_title = r_str_newf ("Top chunk @ 0x%"PFMT64x"\n", main_arena->top);
+
+	if (!cnk || !prev_c) {
+		if (cnk) free(cnk);
+		if (prev_c) free (prev_c);
+		eprintf ("Warning: out of memory\n");
+		return;
+	}
+
+	if (!core || !core->dbg || !core->dbg->maps) {
+		return;
+	}
+
+	r_debug_map_sync (core->dbg);
+	r_list_foreach (core->dbg->maps, iter, map) {
+		if (strstr (map->name, "heap")) {
```

Comment:
put the logic to identify the section in a separate function to avoid the foreach every time you have to use it. for example:

RDebugMap *heap = findHeapInMaps (core->dbg);

also, strstr("heap") can be a weak way to identify that.